src/dymoprint/labeler.py

```diff
@@ -27,28 +29,46 @@ class DymoLabeler:
     _SYN = 0x16
     _MAX_BYTES_PER_LINE = 8  # 64 pixels on a 12mm tape
 
-    def __init__(self, dev):
+    def __init__(self, devout, devin):
         """Initialize the LabelManager object. (HLF)"""
 
         self.cmd = []
         self.response = False
         self.bytesPerLine_ = None
         self.dotTab_ = 0
-        self.dev = open(dev, "rb+")
         self.maxLines = 200
+        self.devout = devout
+        self.devin = devin
 
     def sendCommand(self):
         """Send the already built command to the LabelManager. (MLF)"""
 
         if len(self.cmd) == 0:
             return
-        cmdBin = array.array("B", self.cmd)
-        cmdBin.tofile(self.dev)
+            
+        while len(self.cmd) > 0:
+            synCount = 0
+            pos = -1
+            while synCount < SYNWAIT:
+                try:
+                    pos += self.cmd[pos+1:].index(SYN) + 1
+                except ValueError:
+                    pos = len(self.cmd)
+                    break
+                synCount += 1
+            cmdBin = array.array('B', [ESC, ord('A')])
+            cmdBin.tofile(self.devout)
+            rspBin = self.devin.read(8)
+            rsp = array.array('B', rspBin).tolist()
+            cmdBin = array.array('B', self.cmd[:pos])
+            cmdBin.tofile(self.devout)
+            self.cmd = self.cmd[pos:]
```

Comment:
I'm struggling quite a bit to understand exactly what's going on here.

Could you explain briefly something about the roles of `pos` and `synCount`, and what is the goal of this block? This should somehow relate to the timeouts problem, right? (Sorry, I'm a bit lost here.)

Reply:
Yes it's related to timeout problem. Without it I've managed to print labels but was getting timeout exception. Didn't have time to setup proper dev env so don't know exactly how it works. Tis code is from original dymoprint code. To some extend explanation is in the original code comment:

```python
    _ESC = 0x1b
    _SYN = 0x16
    # Number of commands to send before waiting for a response. This helps
    # to avoid timeouts due to differences between data transfer and
    # printer speeds. I added this because I kept getting "IOError: [Errno
    # 110] Connection timed out" with long labels. Using dev.default_timeout
    # (1000) and the transfer speeds available in the descriptors somewhere, a
    # sensible timeout can also be calculated dynamically.
    _SYNWAIT = 64
```

_ESC and _ESC i guess are just specific command codes that are used to guestimate print time.

Reply:
Ah, I see! I was hoping that you wrote it and could explain it. 

It's not all that complicated, so I'm sure we can figure out what's going on with a bit of effort. I get the impression that whatever is happening here could be done much more straightforwardly with modern Python.

But ok, I'm just really excited that it works! :smiley:

Reply:
I agree :) in fact there are also included 0x1001 and 0x1002 device id's in the original script. I wonder if it's the same LabelManager PnP or an additional device that could be 'officially supported' :)
Also I'll be able to test one more Dymo device tomorrow - different model.

Reply:
0x1001 and 0x1002 are pre- and post- usb-mode-switch of the PnP, so same device but in either storage or printer mode.

Reply:
I figured out how it works and documented it in https://github.com/computerlyrik/dymoprint/commit/47195b01b28ec4d768ec4b9871447a22735b78f4